ENH: Add mode="outlines"

larsoner opened the pull request:
Closes #10690

I have an MRI that is rotated about 15 degrees from canonical which made me want to add this `coord_frame='mri_rotated'`, which just takes the MRI coordinate frame and rotates it to canonical orientation. So for dipoles in head coordinates like:
```
trans = mne.coreg.estimate_head_mri_t(subject)
pos = [[0, 0, 0.05], [0, 0.05, 0], [0.05, 0, 0]]
ori = [[0, 1 / np.sqrt(2), 1 / np.sqrt(2)], [1, 0, 0], [0, 0, 1]]
t = np.arange(len(pos))
amp = [1e-9] * len(pos)
gof = [90.] * len(pos)
dip = mne.Dipole(t, pos, amp, ori=ori, gof=gof)
color = ['r', 'g', 'b']
dip.plot_locations(subject=subject, trans=trans, subjects_dir=subjects_dir,
                   color=color, mode='outlines', coord_frame='mri')
```
Plotting with `coord_frame='mri'` produces this:

![Screenshot from 2022-06-01 17-16-16](https://user-images.githubusercontent.com/2365790/171503311-64bfccd2-baf6-4f6b-8761-3231b3db3df5.png)

While `coord_frame='mri_rotated'` gives this:

![Screenshot from 2022-06-01 17-16-33](https://user-images.githubusercontent.com/2365790/171503320-1ffcd8b0-a49d-486b-b7bd-bcc5a25792f6.png)

Instead we could consider adding support for `coord_frame='mni'` or something instead, but this becomes a bit trickier because the `talxfm` is a full affine, not rigid, so things like `ori` get a bit wacky. But maybe it's better to use that... I can try if people think it makes sense. Maybe eventually both would be useful.

FWIW, in `mne.viz.Brain` we automatically do this rotation, but just make it so that the `view` is always relative to this. The coordinates in 3-space don't have to change because we can just do the right thing with the camera view.

hoechenberger commented:
ooooh this looks nice!!
Only the arrow … stem (? is that what it's called?) could be a little bit narrower, don't you agree?

drammock commented:
> the arrow … stem (? is that what it's called?)

I believe "shaft" is the word you seek

larsoner commented:
I chose the current width to aid readability -- I tried some smaller ones and they were harder to see. We could make it a parameter, too.

agramfort commented:
this ![image](https://output.circle-artifacts.com/output/job/357843dd-2649-47a1-93e9-f6f884cbf546/artifacts/0/dev/_images/sphx_glr_50_background_freesurfer_mne_009.png)

in https://output.circle-artifacts.com/output/job/357843dd-2649-47a1-93e9-f6f884cbf546/artifacts/0/dev/auto_tutorials/forward/50_background_freesurfer_mne.html#cortical-alignment-spherical

but it's likely to be unrelated.

This looks good !

https://output.circle-artifacts.com/output/job/357843dd-2649-47a1-93e9-f6f884cbf546/artifacts/0/dev/auto_tutorials/inverse/20_dipole_fit.html#source-localization-with-equivalent-current-dipole-ecd-fit

larsoner commented:
> but it's likely to be unrelated.

Indeed it's happening on `main` as well

@hoechenberger I tweaked the width to be smaller, added it as a parameter, and added a dot to indicate the arrow pivot (since it differs from the orthoview behavior), better?

![Screenshot from 2022-06-02 12-54-02](https://user-images.githubusercontent.com/2365790/171683594-8c6dc032-27ab-43f9-b254-203b082c3b42.png)

hoechenberger commented:
@larsoner thabks, will take a Look at this soon 

I had a brief discussion with @agramfort this afternoon where I suggested it could be could to have a 4-panel viz with the 3 contours plus the 3D orthoview

Alex was not convinced and said he believes the use cases for both are probably different 

Just wanted to hear what you think?

larsoner commented:
Agreed they're generally different -- the new function is probably better at showing multiple dipoles, but gives less information about the MRI slices/regions.

It's also hopefully easy enough to accomplish by calling each function with appropriate `ax` args after constructing the 4-panel figure however you want to do it.

mmagnuski reviewed (commented):
Great addition! I left just two small comments.

larsoner commented on mne/coreg.py:
```diff
@@ -297,21 +297,7 @@ def _decimate_points(pts, res=10):
-    trans : array, shape = (4, 4)
-        Transformation matrix.
-    """
+    """Convert transformation parameters into a transformation matrix."""
```
Private docstrings almost never stay up to date -- we don't review them carefully (onerous) and we also don't have any automated checking for them. Over time I've been removing these in PRs, and most of the time there is something wrong with them. In this case, the boolean parameter order was wrong for the first argument. We could try to make it so that all private functions are documented 100% correctly, but in the end this might slow us down more than help us out 🤷

mmagnuski replied:
oh, no, I would not suggest that! :)
I just remember that on a few occasions I had to traverse large parts of the code to figure out the inputs to a given private function. So I'm ok with reducing these docstrings, especially if they're out of date as long as we don't remove something very non-obvious that was already present.

hoechenberger replied:
What we really need are type annotations but this is a longer discussion :)

larsoner replied:
Indeed when I remove these I usually try to improve the code commenting if the docstring has something non-obvious.

@hoechenberger feel free to open a separate issue about type annotations if we don't have one already

hoechenberger commented:
Thanks @larsoner! Super cool.
